tests/components/ring/test_init.py

```diff
@@ -386,3 +388,39 @@ async def test_update_unique_id_no_update(
     assert entity_migrated
     assert entity_migrated.unique_id == correct_unique_id
     assert "Fixing non string unique id" not in caplog.text
+
+
+async def test_token_updated(
+    hass: HomeAssistant,
+    mock_ring_devices,
+    freezer: FrozenDateTimeFactory,
+    mock_config_entry: MockConfigEntry,
+) -> None:
+    """Test that the token value is updated in the config entry.
+
+    This simulates the api calling the callback.
+    """
+
+    class mock_client:
+        """Mock client will call the token updater on the second refresh."""
+
+        def __init__(self, auth):
+            self.token_updater = auth.token_updater
+            self.devices = lambda: mock_ring_devices
+            self.async_update_data = AsyncMock()
+            self.async_update_dings = AsyncMock()
+
+        async def async_update_devices(self):
+            self.token_updater({"access_token": "new-mock-token"})
+
+    mock_config_entry.add_to_hass(hass)
+    with patch("homeassistant.components.ring.Ring", new=mock_client):
```

Comment:
Why don't we patch what we need?

Reply:
In the real world calling of the token updater callback happens inside the library if the ring api decides the oath token needs a refresh. To mimic this I need a way to get the actual callback that is passed into the `Auth` class when it's created in `async_setup_entry` and then invoke it during a query:

```py
    auth = Auth(
        f"{APPLICATION_NAME}/{__version__}",
        entry.data[CONF_TOKEN],
        token_updater,
        http_client_session=async_get_clientsession(hass),
    )
    ring = Ring(auth)
```
So to get hold of the callback with just a patch I'd need to patch `__init__` on the `Auth` class. Because the `mock_ring_client` fixture already patches `Auth` I'd have to create the equivalent mock_ring_client again in this test. This works but is possibly more confusing/complicated:
```py
	# This is the same code as in the mock_ring_client fixture except for token updater side_effect
    mock_client = create_autospec(ring_doorbell.Ring)
    mock_client.return_value.devices_data = get_devices_data()
    mock_client.return_value.devices.return_value = mock_ring_devices
    mock_client.return_value.active_alerts.side_effect = get_active_alerts
    mock_client.return_value.async_update_devices.side_effect = lambda: token_updater({"access_token": "new-mock-token"})
	# This is to patch __init__
    token_updater = None
    def _mock_auth_init_(*args, **__):
        nonlocal token_updater
        token_updater = args[2]

    with (
        patch("homeassistant.components.ring.Ring", new=mock_client),
        patch("homeassistant.components.ring.Auth.__init__", side_effect=_mock_auth_init_)
    ):
        assert await hass.config_entries.async_setup(mock_config_entry.entry_id)
	# rest of the test continues as is
```
Can you think of a better way or do you prefer the solution above? (I don't want to patch the `token_updater` function itself as that's what is being tested)

Reply:
Now fixed to use the `call_args` from the `Auth` fixture.